stan/math/prim/prob/wiener4_lcdf.hpp

```diff
@@ -0,0 +1,762 @@
+#ifndef STAN_MATH_PRIM_PROB_WIENER4_LCDF_HPP
+#define STAN_MATH_PRIM_PROB_WIENER4_LCDF_HPP
+
+#include <stan/math/prim/prob/wiener5_lpdf.hpp>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Helper function. Calculate \f$\mathbb{e}^{x}\ if \f$x\ is not too large.
+ *
+ * @param x A scalar
+ * @return \f$\mathbb{e}^{x}\
+ */
+template <typename T_x>
+inline auto rexp(T_x&& x) noexcept {
+  return (x <= 700) ? exp(x) : exp(700);
+}
+
+/**
+ * Helper function. Log of Mill's ratio for the normal distribution
+ *
+ * @param x A scalar
+ * @return Log of Mill's ratio
+ */
+template <typename T_x>
+inline auto logMill(T_x&& x) noexcept {
+  if (x > 1.0e5) {
+    return -log(x);
+  }
+  const auto m = std_normal_lcdf(-x) - std_normal_log(x);
+  return m;
+}
+
+/**
+ * Calculate the probability term 'P' on log scale for distribution
+ *
+ * @param a The boundary separation
+ * @param v The drift rate
+ * @param w The relative starting point
+ * @return 'P' term
+ */
+template <typename T_a, typename T_w, typename T_v>
+inline auto log_probability_distribution(const T_a& a, const T_v& v,
+                                         const T_w& w) noexcept {
+  using ret_t = return_type_t<T_a, T_w, T_v>;
+  auto nearly_one = ret_t(1.0 - 1.0e-6);
+  if (fabs(v) == 0.0) {
+    return ret_t(log1p(-w));
+  }
+  auto minus_two_va_one_minus_w = (-2.0 * v * a * (1.0 - w));
+  ret_t prob;
+  if (minus_two_va_one_minus_w < 0) {
+    const auto exp_arg = exp(minus_two_va_one_minus_w);
+    if (exp_arg >= nearly_one) {
+      return ret_t(log1p(-w));
+    }
+    auto two_vaw = 2 * v * a * w;
+    if (two_vaw > minus_two_va_one_minus_w) {
+      prob = log1p(-exp_arg) - log_diff_exp(two_vaw, minus_two_va_one_minus_w);
+    } else if (two_vaw < minus_two_va_one_minus_w) {
+      prob = log1p(-exp_arg) - log_diff_exp(minus_two_va_one_minus_w, two_vaw);
+    } else {
+      prob = log1p(-exp_arg) - NEGATIVE_INFTY;
+    }
+  } else {
+    const auto exp_arg = exp(-minus_two_va_one_minus_w);
+    if (exp_arg >= nearly_one)
+      return ret_t(log1p(-w));
+    prob = log1p(-exp_arg) - log1p(-exp(2 * v * a));
+  }
+  return prob;
```

Comment:
Early returns are preferred to show when a branch has finished

```suggestion
  auto minus_two_va_one_minus_w = (-2.0 * v * a * (1.0 - w));
  if (minus_two_va_one_minus_w < 0) {
    const auto exp_arg = exp(minus_two_va_one_minus_w);
    if (two_vaw > minus_two_va_one_minus_w) {
      return ret_t(log1p(-exp_arg) - log_diff_exp(two_vaw, minus_two_va_one_minus_w));
    } else if (two_vaw < minus_two_va_one_minus_w) {
      return ret_t(log1p(-exp_arg) - log_diff_exp(minus_two_va_one_minus_w, two_vaw));
    } else {
      return ret_t(log1p(-exp_arg) - NEGATIVE_INFTY);
    }
  } else {
    const auto exp_arg = exp(-minus_two_va_one_minus_w);
    return ret_t(log1p(-exp_arg) - log1p(-exp(2 * v * a)));
  }
```

I also removed `exp_arg >= nearly_one`. Can you show me an example where removing that causes the wrong answer?

Reply:
> Can you show me an example where removing that causes the wrong answer?

In this case, all tests run successfully when I remove `nearly_one`. This is not the case for the other `nealry_one` below.

Reply:
When I replace above in line 150 the 1e-6 with `std::numeric_limits<ret_t>::min()` also an error occurs:

![image](https://github.com/user-attachments/assets/e0fb7f3e-43a7-4f80-9ce8-1be35e879cfc)

Reply:
Does this function give the smallest positive number representable or does it give a negative number? `std::numeric_limits<ret_t>::min()` 
We would need a number neer to zero.
So that the function `log1p(x)` does not become `infinity` when the argument is too near to 1. Instead, if our argument is very near to 1, we want to return `-w`.

Reply:
https://godbolt.org/z/jnrb5Ev8z

It gives back the smallest representable number that is greater than 0